### ISSUE TITLE
Computing a TFIDF vector for a new sentence or phrase

### ISSUE THREAD

**Issue by Fthi-S3074641** [accord-net/framework#813]
Here is a scenario which is based on the example code provided in the [official TFIDF description].(http://accord-framework.net/docs/html/T_Accord_MachineLearning_TFIDF.htm)

I have 3 documents each with 100 words. I create a TFIDF codebook which learns all the documents and ends up with 300 words or less if there are common words. I use the transform function to generate a TFIDF vector which is always a fixed size.

Now, every time I use the Transform() function in a sentence(i.e. an array of tokens), I receive a sparse vector. The size of the sparse vector is the same as the number of words on the codebook. In the scenario I described above the length will be 300. If I work with a large number of documents and big documents the resulting vector gets out of hand.

My question here is to find out if there is a way to specify the length of the vector. In other words a different approach of computing a TFIDF for a sentence. For instance, using the number of documents instead?....

**Comment by cesarsouza** [comment]
Hi @Fthi-S3074641,

Thanks a lot for opening the issue! When you mention about sizes of sparse vectors, do you actually mean that the sparse vectors contain 300 non-zero elements within them? Or do you mean that the .Length property of the sparse vector mentions it contains 300 words?

If you are using sparse vectors, then while the .Length of those vectors can be 300, the actual number of non-zero elements in the sparse vectors should be much lesser unless your sentence happened to contain all those 300 words within it. You can check the number of non-sparse elements by checking the length of the [.Values](http://accord-framework.net/docs/html/P_Accord_Math_Sparse_1_Values.htm) or .Indices properties of the [Sparse vector class](http://accord-framework.net/docs/html/T_Accord_Math_Sparse_1.htm).

If the number of non-zero elements is indeed much lower than 300, then even though the length of the sparse vector can get out of hand, it should not affect learning algorithms that can learn directly from sparse representations such as [kernel SVMs](http://accord-framework.net/docs/html/T_Accord_MachineLearning_VectorMachines_MulticlassSupportVectorMachine_2.htm), especially those learned using [averaged stochastic gradient descent](http://accord-framework.net/docs/html/T_Accord_MachineLearning_VectorMachines_Learning_AveragedStochasticGradientDescent_2.htm).

However, if you do indeed have 300 non-zero elements in your sparse vectors, please share a bit of your data so we could investigate.

Thanks a lot!

Regards,
Cesar

**Comment by Fthi-S3074641** [comment]
Hi!,

Thank you very much for the reply.

I only get a non-zero values (in the sparse vector) equivalent to the
number of words in the sentence (i.e. much less than the 300 words). It is
true it is not a problem for the learning algorithms. The issue here is,
speed. It gets slow when you have a large number of example data that you
need to a TFIDF value for.

I will look into Word2Vec maybe it has something useful.

Best, Fthi.

**Comment by cesarsouza** [comment]
If you want, please share a small example project with the documents you are trying to obtain the TF-IDF features from, and we can try to use it as a basis for performance optimization.

I've seen that there are already some word2vec implementations in .NET, such as [NWord2Vec](https://github.com/richorama/NWord2Vec) and [Word2Vec.NET](https://github.com/eabdullin/Word2Vec.Net). Please let me know what you find about those libraries if you happen to experiment with them. Maybe one of them could be integrated in the framework some day.

Regards,
Cesar

**Comment by Fthi-S3074641** [comment]
Hi, I found this easy to use a W2V, .Net implementation [WordeVecTools](https://github.com/tmteam/Word2vec.Tools). I used it to read a [pre-trained](https://github.com/Kyubyong/wordvectors) Word2Vec models. Currently, I can read a .txt and .vec Word2Vec models. I looked at the implementation of the library and I think it is a very neat implementation.
